### PR TITLE
Dev/device drivers

### LINKED ISSUE #28: Update code to use whatever function calibre uses to get the directory where the book should be sent
Right now, we are using our own function to get the directory of the book on the device. This includes modifying it to change certain symbols to _ or removing extra symbols. There has to be a function somewhere in Calibre's code that does this and we should be using that in case the way it's done changes or we missed a case or something.

### LINKED ISSUE #34: Hard disk error
I'm running calibre 2.53 64bit on Windows 7. It runs locally and files are on the same drive my system runs on.

I'll select a book and select Create/Update X-Rays. It will parse the data then this error will always pop up regardless if my kindle is connected or not:

> There is no disk in the drive. Please insert a disk into \Device\Harddisk1\DR1

I found that clicking 'Try Again' or 'Continue' exactly four times and it works.

This is the log I get on the Creating X-Ray Files job:

> Starting job: Creating X-Ray Files 
>  
> 05-31-2016 18:49:39 Night Film - Marisha Pessl 
> 05-31-2016 18:49:39 	Parsing shelfari data... 
> 05-31-2016 18:49:41 	Getting format specific data... 
> 05-31-2016 18:49:41 	Parsing book data... 
> 05-31-2016 18:49:45 	Creating x-ray... 
> 05-31-2016 18:49:46 	Sending x-ray to device... 
> 
> X-Ray Creation: 
> 	Books Completed: 
> 		Night Film - Marisha Pessl: AZW3 
> 
> X-Ray Sending: 
> 	Books Failed: 
> 		Night Film - Marisha Pessl: 
> 			AZW3: The book is not on the device. 

The book is on the device connected.

### PR DESCRIPTION
**Pull request by szarroug3**
Fixes #28:
- uses calibre's builtin functions to find path to books
- moved finding device function to xray_creator for efficiency

Should also fix #34 

Issue: trying to use the plugin while Calibre is still communicating with the device causes an error saying "Error communicating with the device" -- they're probably trying to talk to it at the same time

### CODE REVIEW

**Comment by stoduk** [comment]
Sorry for slowness, busy over here..  Some general comments:

1) Searching
Urgh, goodreads sucks :)  Only being able to search by title *or* author seems mad, when we have both.

As you pointed out this doesn't work well for some books (eg. A Christmas Carol) - in that case just searching by title seems to do a good job.  Could we try moving to that?  There seems to be some magic on the goodreads side to guess what book you were after.  It might be necessary to then check the author in the books found, perhaps returning the first match.

2) Searching by ASIN or ISBN
Goodreads supports searching by either of these IDs, which we can grab from books in many cases - and is less likely to give the problems mentioned in the previous point.

Is there a reason ASIN searching has been discarded?

3) Aliases
Your comments in Gitter made it sound like goodreads added aliases, but shelfari already had them.. So not sure a change was needed.

In the case you gave I don't see why we'd want to expand "Tiny Tim" - as matching Tiny throughout a book is not going to end well.  Having said that an alias could also be a pseudonym rather than a nickname - in which case expansion makes sense as much as it does for the real name (eg. Marge Simpson and Hootie McBoob are both names that we could validly expand).

So I guess we need to decide if aliases are more likely to be pseudonyms or nicknames - I'd believe nicknames are more likely (so expansion doesn't make sense) but that is a feeling based off little hard facts :)

Could be that a user preference is needed to let the user decide..

File: lib/book.py [hunk 1]
```diff
@@ -244,45 +148,41 @@ def _search_shelfari(self, connection, keywords):
...
                 self._status = self.FAIL
-                self._status_message = self.FAILED_COULD_NOT_CONNECT_TO_SHELFARI
+                self._status_message = self.FAILED_COULD_NOT_CONNECT_TO_GOODREADS
                 raise Exception(self._status_message)
         
         # check to make sure there are results
         if 'did not return any results' in response:
             return connection
-        urlsearch = self.SHELFARI_URL_PAT.search(response)
+        urlsearch = self.GOODREADS_URL_PAT.search(response)
```
Comment (stoduk) [inline lib/book.py]: This is a bug - can't work, as it isn't defined (it is in BookSettings).  However, I can't find a way to hit this codepath - even if I clear the goodreads URL in the per-book JSON file, connect a kindle, and do "send to device" (the only path that could come through here) it still doesn't fail.  Looks like the goodreads URL is being found elsewhere.

Of course, if the goodreads lookup failed then we'd hit problems here, but we might have already barfed earlier..

Is this a redundant code path?

File: book_config.py [hunk 2]
```diff
@@ -130,75 +111,43 @@ def __init__(self, db, ids, parent):
     @property
     def book(self):
         return self._book_settings[self._index]
-    
-    def edit_asin(self, val):
-        self.book.asin = val
 
-    def edit_shelfari_url(self, val):
-        http_string = 'http://www.'
+    def edit_goodreads_url(self, val):
```
Comment (stoduk) [inline book_config.py]: What is the aim of this?

I'm not sure there is much benefit in policing URLs to fit a certain format - what if someone enters it without the www. subdomain?  What if there is a beta.goodreads.com?  And so on.  If the URL fetches and parses ok, that should be good enough IMO.

If you think it is worthwhile then there is something that needs fixing here - as it will just do weird things in certain cases (any time you manually edit anything in "https://www.goodrads.com" rather than pasting in a URL).  For instance: if I remove the sub-doman "www" from the URL, I'll end up with "https://www.goodreads.com/goodreads.com/foobar" which is obviously garbage.

File: book_config.py [hunk 3]
```diff
@@ -130,75 +111,43 @@ def __init__(self, db, ids, parent):
...
+        goodreads_string = 'https://www.goodreads.com/'
         index = 0
-        if val[:len(http_string)] != http_string:
+        if val[:len(goodreads_string)] != goodreads_string:
```
Comment (stoduk) [inline book_config.py]: Minor: if we are keeping this then val.startswith(goodreads_string) would be clearer/more pythonic.

File: book_config.py [hunk 4]
```diff
@@ -130,75 +111,43 @@ def __init__(self, db, ids, parent):
...
                             index += 1
                     else:
                         break
                 else:
                     break
-            self.shelfari_url_edit.setText(http_string + val[index:])
-
-        self.book.shelfari_url = val
-
-    def search_for_asin(self):
```
Comment (stoduk) [inline book_config.py]: see discussion elsewhere: should we be keeping ASIN code, so we can do a lookup using it?  Also, potentially adding ISBN lookups..

File: book_config.py [hunk 5]
```diff
@@ -130,75 +111,43 @@ def __init__(self, db, ids, parent):
...
-        if self.asin_edit.text() != '' and self.asin_edit.text() != 'ASIN not found':
-            url = self.book.search_shelfari(self.asin_edit.text())
-        if not url:
-            if self.book._prefs['asin'] != '':
-                url = self.book.search_shelfari(self.book._prefs['asin'])
-        if not url:
-            if self.book.title != 'Unknown' and self.book.author != 'Unknown':
-                url = self.book.search_shelfari(self.book.title_and_author)
+        if self.book.title != 'Unknown' and self.book.author != 'Unknown':
+            url = self.book.search_goodreads(self.book.title_and_author)
```
Comment (stoduk) [inline book_config.py]: As commented elsewhere: for me searching by title (ie. entering title, and selecting the "title" radio button) gives good results, better than doing a search by title+author using "all" radio button.

File: lib/book.py [hunk 6]
```diff
@@ -125,117 +112,34 @@ def formats_not_failing(self):
...
 
-        self._asin = self._book_settings.prefs['asin'] if self._book_settings.prefs['asin'] != '' else None
-        if not self._asin:
-            identifiers = self._db.field_for('identifiers', self._book_id)
-            self._asin = self._db.field_for('identifiers', self._book_id)['mobi-asin'].decode('ascii') if 'mobi-asin' in identifiers.keys() else None
-            if self._asin:
-                self._book_settings.prefs['asin'] = self._asin
-
-        self._shelfari_url = self._book_settings.prefs['shelfari_url'] if self._book_settings.prefs['shelfari_url'] != '' else None
+        self._author_list = list(self._author) if isinstance(self._author, tuple) else self._author
```
Comment (stoduk) [inline lib/book.py]: Not sure if this makes sense a/o works, it might be fine but just confusing me!  I'll comment what I find odd, and you can hopefully say if it does what is intended :)  Perhaps this is just a need for commenting what the DB returns and what we want out..

self._author is set by a DB lookup.  If we have a single author does it return a string or a tuple/list/whatever containing a single string?  I assume in the multi-author case it is a tuple/list/whatever containing the multiple strings.

I assume it has to be returning a tuple-ish item always - otherwise the " & ".join() won't work well for single authors (we'd be getting 'f & o & o & b & a & r' type nonsense).

So depending on what db.field_for("author") returns we'll have the following, I think:
- returns empty tuple -> self._author is empty tuple, but we'll bail out already
- returns empty string -> self._author is empty string, but we'll bail out already
- returns single author as string -> self._author is nonsense, and self._author_list will be the same
- returns a tuple containing author strings -> self._author is "singleauthor", "author1 & author2", etc. and self_author_list is the same (as it won't be a tuple).

I suspect we'll always get a tuple from the DB lookup, in which case we'll bail early if it is zero length (even possible?) and otherwise self._author will be the "author1 & author2" but self._author_list will be the same.  I imagine you wanted self._author_list to be a list - though IMO it should remain a tuple as we shouldn't be modifying it (the whole difference between list and tuple).

Ah - I now see where this is used - in xray_creator.py where we do "if author not in b.author_list".  author_list is strangely not a list but an "author1 & author2" type string, but due to the magic of python this will work, albeit with corner cases you might not expect..

TL;DR: I think DB returns a tuple,  self._author_list should be set to that same tuple (possibly renaming to self._authors for clarity), self._author should be set afterwards to "soloauthor" or "author1 & author2" as you are currently doing (possibly renaming a self._author_string or similar for clarity).

Phew!

File: lib/book.py [hunk 7]
```diff
@@ -443,28 +281,27 @@ def create_xray(self, aConnection, sConnection, info, log=None, abort=None, remo
             return
         if log: log('%s \t\tCreating x-ray...' % datetime.now().strftime('%m-%d-%Y %H:%M:%S'))
         self._write_xray(info, remove_files_from_dir=remove_files_from_dir)
-        return (aConnection, sConnection)
+        return connection
```
Comment (stoduk) [inline lib/book.py]: I've never found this connection handling very clear, but maybe there is a reason for it.  It seems odd/unpythonic to pass a variable into a function and then pass out what is possibly the same.  Should probably be set on some object that all these functions can see.

Not a big deal, and not changed by this, just pointing out something a bit odd..

Comment (szarroug3) [inline lib/book.py]: The reason for this was that if the connection wasn't used for a long enough time, it would stop working and we'd get an error so somewhere along the way, the connection gets closed and reopened. I wasn't sure if the original variable that was sent in would be updated outside of the function that reopened it so I send it back to be safe. My original intention was to test it but I never got around to it.

File: lib/book.py (hunk 1 above)
Comment (szarroug3) [inline lib/book.py]: That variable is never used. I must've accidentally just replaced the "SHELFARI" part of it and never saw it after. Will remove in next

File: book_config.py (hunk 2 above)
Comment (szarroug3) [inline book_config.py]: I was trying to force the https part since our HTTPSConnection only accepts https pages. I figured I would add the www.goodreads.com part so that people don't try to enter in pages that aren't goodreads (in case they don't know about shelfari being gone or something). You're right though. This needs some work.

File: lib/book.py (hunk 6 above)
Comment (szarroug3) [inline lib/book.py]: Yup. This was messy to figure out. The db returns a string if it's a single author and a tuple if it has multiple authors. The reason I changed it to a list was that at first in xray_creator, I was doing "if author != b.author_list" -- this wouldn't work if one was a list and the other a tuple. It turns out that for some reason, getting the author data from the book on the device doesn't always return the full list of authors. I think we should investigate that some more. Maybe I was doing something wrong -- maybe I uploaded the book with only one author then modified it in Calibre after or something.

**Comment by szarroug3** [comment]
1) Searching
Yeah it does really suck. There has to be a better way to do this. I like the idea of searching by title then checking the authors -- in the case of multiple books with the same name by different authors.

2) Searching by ASIN or ISBN
The only way to get the ASIN is from Amazon. I saw that many people on MobileRead were complaining about the Amazon page not being found. After investigating, I found that some Amazon pages redirect you sometimes not only to a different page but to a different protocol (HTTPS or HTTP). HTTPConnection and HTTPSConnection don't handle that at all. They basically just tell you that you've been redirected and then you have to handle it yourself -- in the case of redirecting to a different protocol, we'd have to open a new connection altogether. I really want to keep keep-alive connections since I'm sure there are people who use this plugin on a huge list of books at once and having to open a connection everytime is very slow. I haven't been able to find any other good way to do keep-alive connections but we can look into this again. 

3) Aliases
I said that because we weren't getting aliases directly from Shelfari as we are from Goodreads. This could affect #37 so I thought it was worth mentioning as it could affect any code you've already written. I find that most aliases are just other names they go by (Vin in Mistborn is called Valette Renoux and Vin Venture a lot). I like the idea of letting the user decide this.

4) Book preferences dialog
Agreed. It is a bit cramped. I LOVE replacing the url edit box with something more useful. I think the title and author is probably sufficient. Maybe even add the book's description (that could get cramped so maybe just a portion of it?)

5) Lower Case Names
Yeah I don't like it either but it didn't bother me too much. I was mostly worried about getting a working version out and fixing little things like this in the next version. 



I'll add issues for each other the things we need to fix/change.

File: lib/book.py (hunk 7 above)
Comment (szarroug3) [inline lib/book.py]: I tested this today. Looks like the original connection will not be updated if the new one isn't returned.

File: book_config.py [hunk 8]
```diff
@@ -32,54 +31,36 @@ def __init__(self, db, ids, parent):
             self._proxy = True
             self._http_address = ':'.join(http_proxy.split(':')[:-1])
```
Comment (stoduk) [inline book_config.py]: Just spotted a problem here - http_proxy is looking up the http proxy, not the https proxy.  I assume they can be different as we are looking http up explicitly.
